### PR TITLE
Merge tab rows with data from websocket

### PR DESCRIPTION
**Pull request by siemiatj**
Just need to add some comments there. Related to #2517

### CODE REVIEW

File: src/reducers/windowHandler.js [hunk 1]
```diff
@@ -421,6 +422,34 @@ export default function windowHandler(state = initialState, action) {
...
+            return data[row.rowId];
+          }
+          return row;
+        });
+
+        // added row
+      } else {
+        rows = reduce(data, (acc, value) => value, []);
+      }
+      let addRowData = Map();
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 20
           

Complexity increasing per file
==============================
- src/reducers/windowHandler.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468&bid=15903067)

File: src/reducers/windowHandler.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import update from 'immutability-helper';
 import { Map, List, Set } from 'immutable';
-import _ from 'lodash';
+import { get, reduce, difference } from 'lodash';
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 3]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
+              });
+            });
+
+            return await Promise.all(requests).then(res => {
+              const changedTabs = {};
+
+              res.forEach(({ data }) => {
+                const rowZero = data[0];
+                const tabId = rowZero.tabId;
+                const rowsById = {};
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 4]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
+
+                data.forEach(row => {
+                  rowsById[row.rowId] = row;
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 5]
```diff
@@ -421,6 +422,34 @@ export default function windowHandler(state = initialState, action) {
...
       };
+    case UPDATE_TAB_ROWS_DATA: {
+      const { data, tabId, scope } = action.payload;
+      const rowData = state[scope].rowData.toJS();
+      let rows = get(rowData, `${tabId}`, []);
+
+      if (rows.length) {
+        rows.map(row => {
+          if (data[row.rowId]) {
+            return data[row.rowId];
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 6]
```diff
@@ -421,6 +422,34 @@ export default function windowHandler(state = initialState, action) {
           rowData: deletedRowData,
         },
       };
+    case UPDATE_TAB_ROWS_DATA: {
+      const { data, tabId, scope } = action.payload;
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 7]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
-              null,
-              null,
-              null
-            )
-          );
+          if (includedTabsInfo) {
+            const requests = [];
+
+            forEach(includedTabsInfo, (tab, tabId) => {
+              const { staleRowIds } = tab;
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 8]
```diff
@@ -421,6 +422,34 @@ export default function windowHandler(state = initialState, action) {
...
+          }
+          return row;
+        });
+
+        // added row
+      } else {
+        rows = reduce(data, (acc, value) => value, []);
+      }
+      let addRowData = Map();
+      addRowData = addRowData.set(tabId, List(rows));
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 9]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
+              res.forEach(({ data }) => {
+                const rowZero = data[0];
+                const tabId = rowZero.tabId;
+                const rowsById = {};
+
+                data.forEach(row => {
+                  rowsById[row.rowId] = row;
+                });
+
+                changedTabs[tabId] = rowsById;
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 10]
```diff
@@ -421,6 +422,34 @@ export default function windowHandler(state = initialState, action) {
...
+        rows.map(row => {
+          if (data[row.rowId]) {
+            return data[row.rowId];
+          }
+          return row;
+        });
+
+        // added row
+      } else {
+        rows = reduce(data, (acc, value) => value, []);
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 11]
```diff
@@ -421,6 +422,34 @@ export default function windowHandler(state = initialState, action) {
           rowData: deletedRowData,
         },
       };
+    case UPDATE_TAB_ROWS_DATA: {
+      const { data, tabId, scope } = action.payload;
+      const rowData = state[scope].rowData.toJS();
+      let rows = get(rowData, `${tabId}`, []);
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 12]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
-              params.windowType,
-              params.docId,
-              null,
-              null,
-              null,
-              null
-            )
-          );
+          if (includedTabsInfo) {
+            const requests = [];
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 13]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
+                    params.docId,
+                    tabId,
+                    rowId
+                  )
+                );
+              });
+            });
+
+            return await Promise.all(requests).then(res => {
+              const changedTabs = {};
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 14]
```diff
@@ -421,6 +422,34 @@ export default function windowHandler(state = initialState, action) {
           rowData: deletedRowData,
         },
       };
+    case UPDATE_TAB_ROWS_DATA: {
+      const { data, tabId, scope } = action.payload;
+      const rowData = state[scope].rowData.toJS();
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 15]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
+
+              res.forEach(({ data }) => {
+                const rowZero = data[0];
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 16]
```diff
@@ -3,8 +3,10 @@ import PropTypes from 'prop-types';
 import React, { Component } from 'react';
 import { connect } from 'react-redux';
 import { push } from 'react-router-redux';
+import { forEach } from 'lodash';
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 17]
```diff
@@ -3,8 +3,10 @@ import PropTypes from 'prop-types';
...
 
 import { addNotification } from '../actions/AppActions';
+import { getData } from '../actions/GenericActions';
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 18]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
+                );
+              });
+            });
+
+            return await Promise.all(requests).then(res => {
+              const changedTabs = {};
+
+              res.forEach(({ data }) => {
+                const rowZero = data[0];
+                const tabId = rowZero.tabId;
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 19]
```diff
@@ -421,6 +422,34 @@ export default function windowHandler(state = initialState, action) {
...
+      let rows = get(rowData, `${tabId}`, []);
+
+      if (rows.length) {
+        rows.map(row => {
+          if (data[row.rowId]) {
+            return data[row.rowId];
+          }
+          return row;
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 20]
```diff
@@ -115,22 +118,62 @@ class MasterWindow extends Component {
...
+
+                data.forEach(row => {
+                  rowsById[row.rowId] = row;
+                });
+
+                changedTabs[tabId] = rowsById;
+              });
+
+              forEach(changedTabs, (changedRows, tabId) => {
+                dispatch(updateTabRowsData('master', tabId, changedRows));
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

**Comment by teosarca** [comment]
@siemiatj  any idea why GET was called twice?
image

**Comment by siemiatj** [comment]
Why stupid Codacy is now complaining about `;` ? Can we finally turn it off as it's only causing noise in the comments ?

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 20
           

Complexity increasing per file
==============================
- src/reducers/windowHandler.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468&bid=15903067)

File: src/reducers/windowHandler.js [hunk 21]
```diff
@@ -421,6 +422,35 @@ export default function windowHandler(state = initialState, action) {
...
+      const rowData = state[scope].rowData.toJS();
+      let rows = get(rowData, `${tabId}`, []);
+
+      // find&replace updated rows (unfortunately it's a table so we'll have to traverse it)
+      if (rows.length) {
+        rows.map(row => {
+          if (data[row.rowId]) {
+            return data[row.rowId];
+          }
+          return row;
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 22]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
+            return await Promise.all(requests).then(res => {
+              const changedTabs = {};
+
+              res.forEach(({ data }) => {
+                const rowZero = data[0];
+                const tabId = rowZero.tabId;
+                const rowsById = {};
+
+                data.forEach(row => {
+                  rowsById[row.rowId] = row;
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 23]
```diff
@@ -421,6 +422,35 @@ export default function windowHandler(state = initialState, action) {
...
+        });
+
+        // added row
+      } else {
+        rows = reduce(data, (acc, value) => value, []);
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 24]
```diff
@@ -421,6 +422,35 @@ export default function windowHandler(state = initialState, action) {
           rowData: deletedRowData,
         },
       };
+    case UPDATE_TAB_ROWS_DATA: {
+      const { data, tabId, scope } = action.payload;
+      const rowData = state[scope].rowData.toJS();
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 25]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
-              null,
-              null
-            )
-          );
+          // some tabs data got updated/row was added
+          if (includedTabsInfo) {
+            const requests = [];
+
+            forEach(includedTabsInfo, (tab, tabId) => {
+              const { staleRowIds } = tab;
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 26]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
-              params.docId,
-              null,
-              null,
-              null,
-              null
-            )
-          );
+          // some tabs data got updated/row was added
+          if (includedTabsInfo) {
+            const requests = [];
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 27]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
+
+                data.forEach(row => {
+                  rowsById[row.rowId] = row;
+                });
+
+                changedTabs[tabId] = rowsById;
+              });
+
+              forEach(changedTabs, (changedRows, tabId) => {
+                dispatch(updateTabRowsData('master', tabId, changedRows));
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js (hunk 2 above)
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 28]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
+              });
+            });
+
+            // wait for all the rows requests to finish
+            return await Promise.all(requests).then(res => {
+              const changedTabs = {};
+
+              res.forEach(({ data }) => {
+                const rowZero = data[0];
+                const tabId = rowZero.tabId;
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 29]
```diff
@@ -421,6 +422,35 @@ export default function windowHandler(state = initialState, action) {
           rowData: deletedRowData,
         },
       };
+    case UPDATE_TAB_ROWS_DATA: {
+      const { data, tabId, scope } = action.payload;
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 30]
```diff
@@ -421,6 +422,35 @@ export default function windowHandler(state = initialState, action) {
...
+      const rowData = state[scope].rowData.toJS();
+      let rows = get(rowData, `${tabId}`, []);
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 31]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
+                const rowsById = {};
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 32]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
+                    tabId,
+                    rowId
+                  )
+                );
+              });
+            });
+
+            // wait for all the rows requests to finish
+            return await Promise.all(requests).then(res => {
+              const changedTabs = {};
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 33]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
+              res.forEach(({ data }) => {
+                const rowZero = data[0];
+                const tabId = rowZero.tabId;
+                const rowsById = {};
+
+                data.forEach(row => {
+                  rowsById[row.rowId] = row;
+                });
+
+                changedTabs[tabId] = rowsById;
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 34]
```diff
@@ -421,6 +422,35 @@ export default function windowHandler(state = initialState, action) {
...
+          }
+          return row;
+        });
+
+        // added row
+      } else {
+        rows = reduce(data, (acc, value) => value, []);
+      }
+      let addRowData = Map();
+      addRowData = addRowData.set(tabId, List(rows));
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js [hunk 35]
```diff
@@ -115,22 +118,66 @@ class MasterWindow extends Component {
...
+                );
+              });
+            });
+
+            // wait for all the rows requests to finish
+            return await Promise.all(requests).then(res => {
+              const changedTabs = {};
+
+              res.forEach(({ data }) => {
+                const rowZero = data[0];
```
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js (hunk 17 above)
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 36]
```diff
@@ -421,6 +422,35 @@ export default function windowHandler(state = initialState, action) {
...
+            return data[row.rowId];
+          }
+          return row;
+        });
+
+        // added row
+      } else {
+        rows = reduce(data, (acc, value) => value, []);
+      }
+      let addRowData = Map();
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/containers/MasterWindow.js (hunk 16 above)
Comment (metas-ts) [inline src/containers/MasterWindow.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

File: src/reducers/windowHandler.js [hunk 37]
```diff
@@ -421,6 +422,35 @@ export default function windowHandler(state = initialState, action) {
...
+    case UPDATE_TAB_ROWS_DATA: {
+      const { data, tabId, scope } = action.payload;
+      const rowData = state[scope].rowData.toJS();
+      let rows = get(rowData, `${tabId}`, []);
+
+      // find&replace updated rows (unfortunately it's a table so we'll have to traverse it)
+      if (rows.length) {
+        rows.map(row => {
+          if (data[row.rowId]) {
+            return data[row.rowId];
```
Comment (metas-ts) [inline src/reducers/windowHandler.js]: Codacy Issue found: [Delete `;`](https://app.codacy.com/manual/metasfresh/metasfresh-webui-frontend/pullRequest?prid=4772468)

**Comment by siemiatj** [comment]
@teosarca no idea relly, I only have one

Screen Shot 2020-01-07 at 10 58 58

**Pull request merged**